### PR TITLE
Explain WAF usage and improve plan for customer domains

### PR DESCRIPTION
**Pull request by jameshochadel**
Related to #2325 

## security considerations
None. Documentation only.

### CODE REVIEW

File: adr/ADR-0009-shield-advanced.md [hunk 1]
```diff
@@ -49,20 +51,24 @@ We will use Terraform in `cg-provision` to do the following:
...
     * Add a [aws_shield_protection_group](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/shield_protection_group) resource so traffic is analyzed across all resources.
 * Change DNS to reference CloudFront distributions instead of the load balancers.
     * Relevant resource type is [aws_route53_record](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/route53_record).
+
+Follow-on work:
 * Configure our LBs to only accept traffic from CloudFront using [custom headers](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-overview.html#forward-custom-headers-restrict-access). Without this step, attackers could make requests to our load balancers directly, bypassing CloudFront and Shield Advanced.
+* See if we could consolidate WAF to commercial only, with one ACL applying to all distributions.
+* Associate customer-brokered CDNs with our protection group.
+* See if we can disable the CDN plan in the External Domain Broker, since the LBs will be protected by a CDN by default. This may not be possible if customers currently customize those distributions.
```
Comment (markdboyd) [inline adr/ADR-0009-shield-advanced.md:67]: I could be very, very wrong, but I don't think disabling the CDN plan in the external domain broker is possible. 

My understanding is that the external domain broker exists to allow cloud.gov customers to route their custom agency domains (e.g. `app.agency.gov`) to their cloud.gov application. To achieve this, ultimately customers need something to point their `app.agency.gov` domain at in their DNS configuration. 

With the external domain broker, we solve this problem by creating a Cloudfront distribution for them. The Cloudfront distribution points at an origin URL for our external domain broker app. And the external domain broker accepts the DNS request and resolves a CNAME to the correct cloud.gov app route for the customer (I'm a bit fuzzy on this). Furthermore, the Cloudfront distribution is bound to an SSL certificate that is specific to the brokered domain (e.g. `app.agency.gov`). 

The bottom line is that if my understanding is correct, then we need the Cloudfront distributions in the external domain broker because they solve the DNS routing problem and certificate management on a per-customer app basis. We couldn't just have our customers point `app.agency.gov` at our platform CDN in front of load balancers because:

- How would it know to route traffic to the correct app?
- How would it handle domain-specific certificates (without attaching them all to a single Cloudfront which seems like a security risk)?

Comment (jameshochadel) [reply adr/ADR-0009-shield-advanced.md:67]: My understanding is different: I think the CDN is an option, but not necessary. 

There are two plans in the broker: One with CDN and one without. If they choose the CDN plan, it works as you described. If they choose the other plan, we just create DNS records that point to their app instead of a CloudFront distribution, and the certificate gets assigned directly to an ALB. (This is why we have 7 ALBs: Each one only has a capacity of twenty-something certs, so we regularly need to increase the count as customers broker additional domains and fill them up.)

Check out the [ALB provision diagram](https://github.com/cloud-gov/external-domain-broker/blob/main/docs/diagrams/alb-provision.png) vs the [CDN provision diagram](https://github.com/cloud-gov/external-domain-broker/blob/main/docs/diagrams/cdn-provision.png) for the full flow. Also [this section](https://github.com/cloud-gov/external-domain-broker#our-solution---cnames-of-cnames) of the readme.

Comment (markdboyd) [reply adr/ADR-0009-shield-advanced.md:67]: OK fair enough. I figured I may have been missing something. But if they do have DNS records pointing straight at the LBs, won't they be bypassing the CDN that we are putting in front of the LBs with Shield Advanced? Also happy to chat about this offline rather than having you explain it piecemeal on GitHub.

Comment (rbogle) [reply adr/ADR-0009-shield-advanced.md:67]: The primary reason for having the CDN plan, I thought, was to provide the global edge based caching that a non CDN domain would lack by default. This is particularly useful for a SPA, static page site, or site with lots of static resources. I  don't think we would want to eliminate that plan in general, but rather add protection (and not caching) to our LBs...

Comment (rbogle) [reply adr/ADR-0009-shield-advanced.md:67]: one more thought: if at some point we need or want to have unterminated encrypted TCP traffic (eg PIV card auth) pass through to the LBs what are the implications? Will custom headers prevent that? Can we require custom headers of another sort for TCP, stand up other LBs just for TCP?

Comment (jameshochadel) [reply adr/ADR-0009-shield-advanced.md:67]: We can eliminate the plan if we choose to have the new platform CDN cover customer domains, and we enable caching. That's possible, and it may save customers time & effort, if our chosen settings work for them and they never need to customize them.

In the case of TCP traffic, I think we'd be using different LBs, specifically Network Load Balancers. ALBs only deal with application-level protocols. I suspect that e.g. PIV traffic would bypass CloudFront entirely; CloudFront seems to be designed for HTTP(S) only. (For instance, its required configuration includes a list of allowed HTTP methods.)

Comment (jameshochadel) [reply adr/ADR-0009-shield-advanced.md:67]: @markdboyd No problem, it's good to have a record. To your question: I think we'll want to reconfigure the external domain broker to start pointing at our CDN, instead of directly at the LBs. So customers will always be using a CDN; it's just a question of which one, ours or "theirs" (brokered).

File: adr/ADR-0009-shield-advanced.md [hunk 2]
```diff
@@ -49,20 +66,24 @@ We will use Terraform in `cg-provision` to do the following:
         * We mostly use ALBs, except for four "classic" ELBs. Three of those are internal.
     * A single distribution can point to [up to 25 origins](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/distribution-overview.html), but seems to be associated with a single domain. So the number of distributions we will need is 1:1 with either the number of load balancers or the number of domains, in the case that multiple load balancers support a single domain.
     * For reference, a single AWS account has a default quota of [200 distributions](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/cloudfront-limits.html#limits-web-distributions).
+* Create Web ACL(s) in the commercial AWS account and associate them with WAF and each CloudFront distribution.
+    * I think we can have one ACL for all distributions.
+    * WAF is already enabled in our GovCloud account. We will most likely keep it enabled so that customers who broker their own CDNs have at least the protection of the Managed Rulesets.
```
Comment (markdboyd) [inline adr/ADR-0009-shield-advanced.md:71]: I don't think Cloudfront is available in GovCloud. So while it's true that WAF is enabled and we're using it in GovCloud, it's not relevant to customers brokering their own CDNs

File: adr/ADR-0009-shield-advanced.md (hunk 1 above)
Comment (markdboyd) [reply adr/ADR-0009-shield-advanced.md:82]: I'm still not sure if we can do this for practical reasons. For example, customers ask us to invalidate their CDN cache periodically. We couldn't do that if we had one CDN for all customers. 

It's a minor point, but not sure if this point is worth including in the ADR

File: adr/ADR-0009-shield-advanced.md [hunk 3]
```diff
@@ -1,22 +1,39 @@
...
+
+We want to mitigate the impact of DDoS attacks, monitor the platform as they occur, and investigate them after they are over. To accomplish this, we want to make several changes to the platform.
+
+1. Enable additional Web Application Firewall (WAF) rules across the entire platform. WAF uses rules to filter incoming requests based on certain patterns.
+2. Enable AWS Config. Config would help us detect if an AWS resource changed during an attack. Using Config is an AWS best practice recommended to us during an AWS security review. This will be covered in a separate ADR.
+3. Enable AWS GuardDuty. GuardDuty monitors our AWS resources for unusual and potentially malicious activity like escalation of privileges and communication with malicious IP addresses. GuardDuty would help us detect if an AWS resource was compromised during an attack. Using GuardDuty is an AWS best practice recommended to us during an AWS security review. This will be covered in a separate ADR.
```
Comment (markdboyd) [inline adr/ADR-0009-shield-advanced.md:11]: I think it's probably just clearer to leave these for the separate ADRs and focus in this ADR on how we're using WAF + Shield Advanced to actually mitigate the attacks

Config + GuardDuty will serve the complementary purpose of monitoring for unusual activity that could be associated with an attack

File: adr/ADR-0009-shield-advanced.md [hunk 4]
```diff
@@ -1,22 +1,39 @@
-# CloudFront and Shield Advanced
+# DDoS Mitigation and Investigation
```
Comment (markdboyd) [inline adr/ADR-0009-shield-advanced.md:1]:
```suggestion
# DDoS Mitigation: WAF and Shield Advanced
```

File: adr/ADR-0009-shield-advanced.md (hunk 2 above)
Comment (jameshochadel) [reply adr/ADR-0009-shield-advanced.md:71]: Alas, I could've been clearer here. Even when we turn on CloudFront + Shield Advanced for the platform, customers who broker their own CDNs will have no Shield Advanced coverage until we do follow-on work to update the External Domain Broker and backfill existing distributions to have protection. So this line is saying that even though we'll have "double WAF" — one in GovCloud and one in Commercial — and that might seem unintuitive at first, we'll want to leave it like this so customers who broker their own CDNs get baseline protection while we work on protecting them with Shield Advanced.

File: adr/ADR-0009-shield-advanced.md (hunk 3 above)
Comment (jameshochadel) [reply adr/ADR-0009-shield-advanced.md:11]: Sure, I can do that.

File: adr/ADR-0009-shield-advanced.md (hunk 2 above)
Comment (markdboyd) [reply adr/ADR-0009-shield-advanced.md:71]: Ah okay. Can we update the document along the lines of your comment then?

Comment (jameshochadel) [reply adr/ADR-0009-shield-advanced.md:71]: Sure thing!

File: adr/ADR-0009-shield-advanced.md (hunk 1 above)
Comment (jameshochadel) [reply adr/ADR-0009-shield-advanced.md:79]: @markdboyd Good point about the invalidation. I'll change this to a new task for documenting the new architecture for customers, and explaining when they may want to broker their own CDN.

**Review by rcgottlieb — APPROVED** [review]

**Review by markdboyd — APPROVED** [review]
